### PR TITLE
SAAS-12945/iam-policies-plugins

### PR DESCRIPTION
**Pull request by alphadev4**

### CODE REVIEW

File: plugins/aws/lambda/lambdaOldRuntimes.js [hunk 1]
```diff
@@ -10,12 +10,24 @@ module.exports = {
     link: 'http://docs.aws.amazon.com/lambda/latest/dg/current-supported-versions.html',
     recommended_action: 'Upgrade the Lambda function runtime to use a more current version.',
     apis: ['Lambda:listFunctions'],
+    settings: {
+        lambda_runtime_fail: {
+            name: 'Lambda Runtime Fail',
+            description: 'Return a failing result for lambda runtime before this number of days for their end of life date.',
+            regex: '^[1-9]{1}[0-9]{0,3}$',
+            default: 30
```
Comment (AkhtarAmir) [inline plugins/aws/lambda/lambdaOldRuntimes.js:18]: lets keep this 0

Comment (alphadev4) [reply plugins/aws/lambda/lambdaOldRuntimes.js:18]: ok

File: plugins/aws/iam/iamRolePolicies.js [hunk 2]
```diff
@@ -69,7 +69,15 @@ module.exports = {
             description: 'Ignores roles that contain the provided tag. Give key-value pair i.e. env:Finance ',
             regex: '^.*$',
             default: ''
-        }
+        },
+        resource_specific_wildcards: {
+            name: 'Service Specific Wildcards',
+            description: 'Resource specific wildcards in attached policies (inline and managed) which matches this regex. '+
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:75]: can we make this more descriptive? like:
input to flag resources based on regex. All policy resource which not pass this regex will be flagged.

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/lambda/lambdaOldRuntimes.js [hunk 3]
```diff
@@ -10,12 +10,24 @@ module.exports = {
...
+        }
+    },
 
     run: function(cache, settings, callback) {
         var results = [];
         var source = {};
         var regions = helpers.regions(settings);
 
+        var config = {
+            lambda_runtime_fail: settings.lambda_runtime_fail || this.settings.lambda_runtime_fail.default
```
Comment (AkhtarAmir) [inline plugins/aws/lambda/lambdaOldRuntimes.js:28]:
```suggestion
            lambda_runtime_fail: parseInt(settings.lambda_runtime_fail || this.settings.lambda_runtime_fail.default)
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 4]
```diff
@@ -69,7 +69,32 @@ module.exports = {
             description: 'Ignores roles that contain the provided tag. Give key-value pair i.e. env:Finance ',
             regex: '^.*$',
             default: ''
-        }
+        },
+        iam_policy_resource_specific_wildcards: {
+            name: 'Resource Specific Wildcards',
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:74]:
```suggestion
            name: 'IAM Policy Resource Specific Wildcards',
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 5]
```diff
@@ -69,7 +69,32 @@ module.exports = {
...
+                                "s3:GetObject"
+                            ],
+                            "Resource": [
+                                "arn:aws:s3:::*"
+                            ]
+                        }
+                ]
+            }` +
+                'If iam_policy_resource_specific_wildcards is set to following regex to ignore resource wildcards /^[a-z]+:[a-z]+:[a-z0-9]+:::[a-z]+$/, FAIL result will be generated. ' +
+                'If iam_policy_resource_specific_wildcards is set to default value, PASS result will be generated.',
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:93]:
```suggestion
            description: 'Allows policy resources to flag based on regular expression. All the resources in IAM policy, inline or managed, will be tested against this regex and if they don't pass the regex, they will be flagged by the plugin.',
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 6]
```diff
@@ -79,14 +104,17 @@ module.exports = {
...
+            iam_policy_resource_specific_wildcards: settings.iam_policy_resource_specific_wildcards || this.settings.iam_policy_resource_specific_wildcards.default
         };
 
         config.ignore_service_specific_wildcards = (config.ignore_service_specific_wildcards === 'true');
         config.ignore_identity_federation_roles = (config.ignore_identity_federation_roles === 'true');
         config.ignore_aws_managed_iam_policies = (config.ignore_aws_managed_iam_policies === 'true');
         config.ignore_customer_managed_iam_policies = (config.ignore_customer_managed_iam_policies === 'true');
 
+        var allowRegex = (config.iam_policy_resource_specific_wildcards &&
+            config.iam_policy_resource_specific_wildcards.length) ? new RegExp(config.iam_policy_resource_specific_wildcards) : false;
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:117]:
```suggestion
        var allowedRegex = RegExp(config.iam_policy_resource_specific_wildcards);
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 7]
```diff
@@ -206,9 +234,10 @@ module.exports = {
                             getPolicyVersion.data.PolicyVersion.Document) {
                             let statements = helpers.normalizePolicyDocument(
                                 getPolicyVersion.data.PolicyVersion.Document);
+
                             if (!statements) break;
 
-                            addRoleFailures(roleFailures, statements, 'managed', config.ignore_service_specific_wildcards);
+                            addRoleFailures(roleFailures, statements, 'managed', config.ignore_service_specific_wildcards, allowRegex);
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:240]:
```suggestion
                            addRoleFailures(roleFailures, statements, 'managed', config.ignore_service_specific_wildcards, allowedRegex);
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 8]
```diff
@@ -224,10 +253,11 @@ module.exports = {
                         getRolePolicy[policyName] &&
                         getRolePolicy[policyName].data &&
                         getRolePolicy[policyName].data.PolicyDocument) {
+
                         var statements = helpers.normalizePolicyDocument(
                             getRolePolicy[policyName].data.PolicyDocument);
                         if (!statements) break;
-                        addRoleFailures(roleFailures, statements, 'inline', config.ignore_service_specific_wildcards);
+                        addRoleFailures(roleFailures, statements, 'inline', config.ignore_service_specific_wildcards, allowRegex);
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:260]:
```suggestion
                        addRoleFailures(roleFailures, statements, 'inline', config.ignore_service_specific_wildcards, allowedRegex);
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 9]
```diff
@@ -270,9 +302,18 @@ function addRoleFailures(roleFailures, statements, policyType, ignoreServiceSpec
                     }
                 }
                 if (wildcards.length) failMsg = `Role ${policyType} policy allows wildcard actions: ${wildcards.join(', ')}`;
+            } else if (regResource && statement.Resource && statement.Resource.length) {
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:305]:
```suggestion
            }
            if (regResource && statement.Resource && statement.Resource.length) {
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 10]
```diff
@@ -270,9 +302,18 @@ function addRoleFailures(roleFailures, statements, policyType, ignoreServiceSpec
...
+                // Check each resource for wildcard
+                let wildcards = [];
+                for (var resource of statement.Resource) {
+                    if (!regResource.test(resource)) {
+                        wildcards.push(resource);
+                    }
+                }
+                if (wildcards.length) failMsg = `Role ${policyType} policy allows resource wildcards: ${wildcards.join(', ')}`;
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamRolePolicies.js:313]:
```suggestion
                if (wildcards.length) failMsg = `Role ${policyType} policy does not match provided regex: ${wildcards.join(', ')}`;
```

File: plugins/aws/iam/iamRolePolicies.spec.js [hunk 11]
```diff
@@ -296,7 +373,7 @@ describe('iamRolePolicies', function () {
 
         it('should PASS if role policy allows wildcard actions but ignore service specific roles setting is enabled', function (done) {
             const cache = createCache([listRoles[0]],getRole[0], listAttachedRolePolicies[2], null, null, getPolicy[0], getPolicyVersion[0]);
-            iamRolePolicies.run(cache, { ignore_service_specific_wildcards: 'true' }, (err, results) => {
+            iamRolePolicies.run(cache, { ignore_service_specific_wildcards: 'true', resource_specific_wildcards: 'true' }, (err, results) => {
```
Comment (alphadev4) [inline plugins/aws/iam/iamRolePolicies.spec.js:376]:
```suggestion
            iamRolePolicies.run(cache, {resource_specific_wildcards: 'true' }, (err, results) => {
```

File: plugins/aws/iam/iamRolePolicies.spec.js [hunk 12]
```diff
@@ -358,5 +435,16 @@ describe('iamRolePolicies', function () {
                 done();
             });
         });
+       
+        it('should FAIL if role policy allows resources which does not match regex in iam_policy_resource_specific_wildcards', function (done) {
+            const cache = createCache([listRoles[2]],getRole[1], listAttachedRolePolicies[3], null, null, getPolicy[1], getPolicyVersion[1]);
+            iamRolePolicies.run(cache,{ ignore_service_specific_wildcards: 'true',iam_policy_resource_specific_wildcards: '^[a-z]+:[a-z]+:[a-z0-9]+:::[a-z]+$'}, (err, results) => {
```
Comment (alphadev4) [inline plugins/aws/iam/iamRolePolicies.spec.js:441]:
```suggestion
            iamRolePolicies.run(cache, {iam_policy_resource_specific_wildcards: '^[a-z]+:[a-z]+:[a-z0-9]+:::[a-z]+$'}, (err, results) => {
```

File: plugins/aws/iam/rolePolicyUnusedServices.js [hunk 13]
```diff
@@ -75,6 +75,36 @@ module.exports = {
...
+                            "Resource": [
+                                "arn:aws:s3:::*"
+                            ]
+                        }
+                ]
+            }` +
+                'If iam_policy_resource_specific_wildcards is set to following regex to ignore resource wildcards /^[a-z]+:[a-z]+:[a-z0-9]+:::[a-z]+$/, FAIL result will be generated. ' +
+                'If iam_policy_resource_specific_wildcards is set to default value, PASS result will be generated.',
+            regex: '^.*$',
+            default: '^.*$',
```
Comment (alphadev4) [inline plugins/aws/iam/rolePolicyUnusedServices.js:107]:
```suggestion
        whitelist_unused_actions_for_resources: {
            name: 'Whitelist Unused Actions For Resources',
            description: 'A comma-separated list indicating which actions for specific service should be ignored i.e. Ignore the following actions ec2:CancelReservedInstancesListing, ec2:AcceptVpcPeeringConnection for ec2 Service',
            regex: '.*$',
            default: ''
        },
        iam_policy_resource_specific_wildcards: {
            name: 'IAM Policy Resource Specific Wildcards',
            description: 'Allows policy resources to flag based on regular expression. All the resources in IAM policy, inline or managed, will be tested against this regex and if they don\'t pass the regex, they will be flagged by the plugin.',
            regex: '^.*$',
            default: '^.*$',
```

File: plugins/aws/iam/rolePolicyUnusedServices.js [hunk 14]
```diff
@@ -86,14 +116,19 @@ module.exports = {
...
         };
 
         config.ignore_service_specific_wildcards = (config.ignore_service_specific_wildcards === 'true');
         config.ignore_identity_federation_roles = (config.ignore_identity_federation_roles === 'true');
         config.ignore_aws_managed_iam_policies = (config.ignore_aws_managed_iam_policies === 'true');
         config.ignore_customer_managed_iam_policies = (config.ignore_customer_managed_iam_policies === 'true');
         config.whitelist_unused_services = config.whitelist_unused_services.replace(/\s/g, '');
+        config.whitelist_unused_actions_for_resources = config.whitelist_unused_actions_for_resources.replace(/\s/g, '').toLowerCase();
+        var allowRegex = (config.iam_policy_resource_specific_wildcards &&
+            config.iam_policy_resource_specific_wildcards.length) ? new RegExp(config.iam_policy_resource_specific_wildcards) : false;
```
Comment (alphadev4) [inline plugins/aws/iam/rolePolicyUnusedServices.js:131]:
```suggestion
        var allowedRegex = new RegExp(config.iam_policy_resource_specific_wildcards);
```

File: plugins/aws/iam/rolePolicyUnusedServices.js [hunk 15]
```diff
@@ -354,7 +389,7 @@ module.exports = {
                                     }
                                 }
 
-                                addRoleFailures(roleFailures, statements, 'managed', config.ignore_service_specific_wildcards);
+                                addRoleFailures(roleFailures, statements, 'managed', config.ignore_service_specific_wildcards, allowRegex);
```
Comment (alphadev4) [inline plugins/aws/iam/rolePolicyUnusedServices.js:392]:
```suggestion
                                addRoleFailures(roleFailures, statements, 'managed', config.ignore_service_specific_wildcards, allowedRegex);
```

File: plugins/aws/iam/rolePolicyUnusedServices.js [hunk 16]
```diff
@@ -402,7 +438,7 @@ module.exports = {
                                 }
                             }
 
-                            addRoleFailures(roleFailures, statements, 'inline', config.ignore_service_specific_wildcards);
+                            addRoleFailures(roleFailures, statements, 'inline', config.ignore_service_specific_wildcards, allowRegex);
```
Comment (alphadev4) [inline plugins/aws/iam/rolePolicyUnusedServices.js:441]:
```suggestion
                            addRoleFailures(roleFailures, statements, 'inline', config.ignore_service_specific_wildcards, allowedRegex);
```

File: plugins/aws/iam/rolePolicyUnusedServices.js [hunk 17]
```diff
@@ -446,6 +484,15 @@ function addRoleFailures(roleFailures, statements, policyType, ignoreServiceSpec
                     }
                 }
                 if (wildcards.length) failMsg = `Role ${policyType} policy allows wildcard actions: ${wildcards.join(', ')}`;
+            } else if (regResource && statement.Resource && statement.Resource.length) {
```
Comment (alphadev4) [inline plugins/aws/iam/rolePolicyUnusedServices.js:487]:
```suggestion
            }
            if (statement.Resource && statement.Resource.length) {
```

File: plugins/aws/iam/iamRolePolicies.spec.js [hunk 18]
```diff
@@ -358,5 +435,16 @@ describe('iamRolePolicies', function () {
...
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('policy allows resource wildcards:');
```
Comment (alphadev4) [inline plugins/aws/iam/iamRolePolicies.spec.js:443]:
```suggestion
                expect(results[0].message).to.include('policy does not match provided regex');
```

File: plugins/aws/iam/rolePolicyUnusedServices.spec.js [hunk 19]
```diff
@@ -529,5 +590,25 @@ describe('rolePolicyUnusedServices', function () {
...
+                expect(results.length).to.equal(1);
+                expect(results[0].message).to.include('Role does not have overly-permissive');
+                expect(results[0].status).to.equal(0);
+                done();
+            });
+        });
+
+        it('should FAIL if role policy allows resources which does not match regex in iam_policy_resource_specific_wildcards', function (done) {
+            const cache = createCache([configStatus[0]], discoveredResources, [listRoles[0]], getRole[0], {}, listRolePolicies[1], getRolePolicy[5]);
+            rolePolicyUnusedServices.run(cache,{ ignore_service_specific_wildcards: 'true',iam_policy_resource_specific_wildcards: '^[a-z]+:[a-z]+:[a-z0-9]+:::[a-z]+$'}, (err, results) => {
```
Comment (alphadev4) [inline plugins/aws/iam/rolePolicyUnusedServices.spec.js:606]:
```suggestion
            rolePolicyUnusedServices.run(cache, {iam_policy_resource_specific_wildcards: '^[a-z]+:[a-z]+:[a-z0-9]+:::[a-z]+$'}, (err, results) => {
```

File: plugins/aws/iam/iamRolePolicies.js [hunk 20]
```diff
@@ -206,9 +215,10 @@ module.exports = {
                             getPolicyVersion.data.PolicyVersion.Document) {
                             let statements = helpers.normalizePolicyDocument(
                                 getPolicyVersion.data.PolicyVersion.Document);
+
```
Comment (alphadev4) [inline plugins/aws/iam/iamRolePolicies.js:218]:
```suggestion
```

File: plugins/aws/iam/rolePolicyUnusedServices.js [hunk 21]
```diff
@@ -437,6 +454,8 @@ function addRoleFailures(roleFailures, statements, policyType, ignoreServiceSpec
                 failMsg = `Role ${policyType} policy allows all actions on all resources`;
             } else if (statement.Action.indexOf('*') > -1) {
                 failMsg = `Role ${policyType} policy allows all actions on selected resources`;
+            } else if (statement.Resource && statement.Resource.indexOf('*') > -1){
```
Comment (alphadev4) [inline plugins/aws/iam/rolePolicyUnusedServices.js:457]:
```suggestion
            } else if (statement.Resource && statement.Resource == '*'){
```

**Pull request merged**